monkey/infection_monkey/puppet/plugin_source_extractor.py

```diff
@@ -0,0 +1,116 @@
+import io
+from pathlib import Path
+from tarfile import TarFile, TarInfo
+
+from common.agent_plugins import AgentPlugin
+from common.utils.file_utils import create_secure_directory
+
+
+class PluginSourceExtractor:
+    """
+    Extracts a plugin's source code archive to disk
+    """
+
+    def __init__(self, plugin_destination_directory: Path):
+        """
+        :param plugin_destination directory: A directory where this component will extract plugin
+                                             source code archives
+        """
+        self._plugin_destination_directory = plugin_destination_directory.resolve()
+
+    @property
+    def plugin_destination_directory(self) -> Path:
+        """
+        The directory where this component will extract plugin source code archives
+        """
+        return self._plugin_destination_directory
+
+    def extract_plugin_source(self, agent_plugin: AgentPlugin):
+        """
+        Extracts an `AgentPlugin`'s source code archive
+
+        The `AgentPlugin`'s source code archive will be extracted to
+        `self.plugin_destination_directory / agent_plugin.manifest.name`. While this method attempts
+        to prevent directory traversal and other relevant vulnerabilities, it is still advisable
+        that you do not use this to extract archives/plugins from untrusted sources.
+
+        :param agent_plugin: An `AgentPlugin` to extract
+        :raises ValueError: If the agent's source code archive is potentially malicious or otherwise
+                            can not be extracted
+        """
+        destination = self._get_plugin_destination_directory(agent_plugin)
+        create_secure_directory(destination)
+
+        archive = TarFile(fileobj=io.BytesIO(agent_plugin.source_archive), mode="r")
+
+        # We check the entire archive to detect any malicious activity **before** we extract any
+        # files. This is a paranoid approach that prevents against partial extraction of malicious
+        # archives; either the whole archive is judged to be safe and extracted, or nothing is
+        # extracted.
+        _detect_malicious_archive(destination, archive)
+        _safe_extract(destination, archive)
```

Comment:
This looks reusable enough to put in common.

Reply:
It's probably reusable enough to be its own package on PyPI. But I'd prefer not to move it anywhere until we want to use it somewhere else.

I need to audit the Island code that extracts tar files and see if it can benefit from this. If it can, I'll create an issue.